### PR TITLE
Update pulumi dockerfile to multiarch

### PR DESCRIPTION
**Pull request by MMartyn**
I would like to setup devcontainers for use by the developers when they are working on my custom provider. Currently since there isn't an arm64 version, the experience is slow and painful on Apple Silicon. This change updates the Dockerfile so that all that is needed is the `--platform=linux/arm64` in the build and it will take care of switching out all the places needed to build. Note: The build matrices will need to be updated if this is something you would consider adding.

Thanks.

### CODE REVIEW

File: docker/pulumi/Dockerfile [hunk 1]
```diff
@@ -36,11 +39,17 @@ RUN apt-get update -y && \
 
 # Install cloud tools
 RUN \
+  # Setup environment variables for architecture-specific packages
+  if [ "$TARGETARCH" = "amd64" ]; then \
+    AWSCLI_ARCH=x86_64; \
+  else \
+    AWSCLI_ARCH=aarch64; \
```
Comment (tgummerer) [inline docker/pulumi/Dockerfile:46]: We should probably switch the default around here, so amd64 is the default?  Or maybe set that as a default above if no TARGETARCH argument is given?

File: docker/pulumi/Dockerfile [hunk 2]
```diff
@@ -166,14 +184,21 @@ ENV GORELEASER_VERSION 1.11.4
 
 SHELL ["/bin/bash", "-o", "errexit", "-o", "nounset", "-o", "pipefail", "-c"]
 
-RUN curl \
+RUN \
+  # Setup environment variables for architecture-specific packages
+  if [ "$TARGETARCH" = "amd64" ]; then \
+   GORELEASER_ARCH=x86_64; \
```
Comment (tgummerer) [inline docker/pulumi/Dockerfile:190]:
```suggestion
    GORELEASER_ARCH=x86_64; \
```

very nitpicky here, but I think this is missing a space.

File: docker/pulumi/Dockerfile [hunk 3]
```diff
@@ -1,12 +1,15 @@
 FROM debian:12 AS base
 
+ARG TARGETARCH
```
Comment (tgummerer) [inline docker/pulumi/Dockerfile:3]: This probably deserves a comment about which values are accepted here (`amd64` and `arm64` after this PR is merged if I'm not mistaken?)

**Review by tgummerer — COMMENTED** [review]
Thanks!  I do think this would be really useful to have.  I left a few comments inline.  Would you be willing to update the build matrix as well?

@julienp is the expert on these docker files, so I would want him to have a look at this as well (he's on PTO, and will be back next week)

File: docker/pulumi/Dockerfile (hunk 3 above)
Comment (MMartyn) [reply docker/pulumi/Dockerfile:3]: The TARGETARCH argument is automatically provided by docker. See: https://docs.docker.com/build/building/variables/#pre-defined-build-arguments
https://docs.docker.com/reference/dockerfile/#automatic-platform-args-in-the-global-scope

I could still make the requested updates if you think it is worth it.

**Comment by MMartyn** [comment]
I believe I addressed all of the comments. Looking at the addition of nonroot, I think I all is well there as well.

**Comment by julienp** [comment]
This looks great! I think we also need to matrix the arch in `ci.yml`. That's only running the tests, so shouldn't need all the work for the docker manifests.

Annoyingly, there's also `sync-ecr.yml` and `sync-ghcr.yml` that push the images to the GitHub and amazon registries, which probably need to recreate the manifests again.

**Comment by MMartyn** [comment]
BTW regarding the build failure, seems like a transient GH issue https://github.com/orgs/community/discussions/139743

**Comment by julienp** [comment]
> Added the code for gcp and aws. Didn't touch the ci one since it looks like all the other ones just use the amd64 version for testing.

I think we do test the language specific images on arm64, but I'll create a separate issue for myself to look into this for the kitchen sink image. There will likely be some issues due to us using qemu to test arm64, I know at lest dotnet can't run under qemu, and there might be some issues with nodejs also.

**Comment by julienp** [comment]
Test failure on this PR are because we aren't setup well for running tests for external contributions. I ran the code changes on this branch here https://github.com/pulumi/pulumi-docker-containers/actions/runs/11035725145/job/30690370889?pr=285

**Review by julienp — APPROVED** [review]
Resolved a merge conflict due to https://github.com/pulumi/pulumi-docker-containers/commit/2a13fe5d0bb3d1099182c98a517306a3bfae50e8, which in hindsight I could have merged this PR instead and it would have fixed the issue also, since it splits the push for the latest tag out to the manifests step!

**Comment by julienp** [comment]
Unfortunately we had to disable the release builds of these images on arm64, because it takes too long to install the Python versions provided by the image when running in Docker + QEMU.

Tracking re-enabling these here https://github.com/pulumi/pulumi-docker-containers/issues/297

In the meantime, the images can be built locally on an arm64 machine.

**Comment by MMartyn** [comment]
Alright, looks like they plan on opening it up to open source by EOY. I went ahead and built the images and pushed them to my ECR so I can get my internal devs the ability to use devcontainers. I'll keep my eye on that #297

**Pull request merged**